### Description: This is a case study about the financial fines paid by doctors working in research dept of a company. 
 

In [1]:
import pandas as pd
import numpy as np


In [25]:
#importing data sets
fine=pd.read_csv(r"C:\Users\MUHAMMED JUNAID N K\Desktop\Untitled Folder\Fines paid  by doctors.csv")



In [26]:
emails=pd.read_excel(r"C:\Users\MUHAMMED JUNAID N K\Desktop\Untitled Folder\Emails sent.xlsx")

### 'FInes paid by doctors' dataset contains the online and cash fines. Some doctors paid both fines in cash as well as online.
 

### 'Email sent' dataset contains the doctor names, the date at which email stating the fines was sent and the office at which the doctors worked. Each row in this dataset represents an email sent to a doctors.

In [27]:
fine.head(10)

org_indiv first_name    last_name          city state  \
0    3-D MEDICAL SERVICES LLC     STEVEN  DEITELZWEIG   NEW ORLEANS    LA   
1            AA DOCTORS, INC.     AAKASH        AHUJA   PASO ROBLES    CA   
2      ABBO, LILIAN MARGARITA     LILIAN         ABBO         MIAMI    FL   
3      ABBO, LILIAN MARGARITA     LILIAN         ABBO         MIAMI    FL   
4      ABBO, LILIAN MARGARITA     LILIAN         ABBO         MIAMI    FL   
5       ABDULLAH RAFFEE MD PC   ABDULLAH       RAFFEE         FLINT    MI   
6             ABEBE, SHEILA Y     SHEILA        ABEBE  INDIANAPOLIS    IN   
7             ABEBE, SHEILA Y     SHEILA        ABEBE  INDIANAPOLIS    IN   
8  ABILENE FAMILY FOOT CENTER      GALEN    ALBRITTON       ABILENE    TX   
9             ABOLNIK, IGOR Z       IGOR      ABOLNIK         PROVO    UT   

                  category  cash_fine   online_fine   
0    Professional Advising  $2,625.00          $-     
1        Expert-Led Forums  $1,000.00          $-     
2  Business Related Travel       $-         $448.00   
3                    Meals       $-         $119.00   
4    Professional Advising  $1,800.00          $-     
5        Expert-Led Forums    $750.00          $-     
6        Educational Items       $-          $47.00   
7        Expert-Led Forums    $825.00          $-     
8    Professional Advising  $3,000.00          $-     
9  Business Related Travel       $-         $396.00

### Q.1  Standardise the mail_sent_date column values to YYYY-MM-DD standard date format


In [33]:
emails['mail_sent_date']=pd.to_datetime(emails['mail_sent_date']).dt.strftime('%Y-%m-%d')

In [71]:
emails.head(10)

name_last name_first name_middle       mail_sent_date  \
0   ADELGLASS    JEFFREY          M.  2009-05-22 00:00:00   
1    ADKINSON         N.    FRANKLIN           2010-04-17   
2       ALLEN       MARK          S.           2012-01-26   
3   AMSTERDAM     DANIEL         NaN           2014.11.15   
4     AMSTUTZ     HARLAN          C.           2014.07.17   
5    ANDERSON         C.      JOSEPH           2010-02-22   
6     ANDREWS      DAVID          W.           2010-07-17   
7        AQEL       RAED         NaN           2012-10-27   
8  ARROWSMITH      PETER          N.           2014.01.18   
9        BARR       JOHN          D.           2010-01-11   

                                         office  
0       Center for Drug Evaluation and Research  
1  Center for Biologics Evaluation and Research  
2    Center for Devices and Radiological Health  
3  Center for Biologics Evaluation and Research  
4    Center for Devices and Radiological Health  
5    Center for Devices and Radiological Health  
6  Center for Biologics Evaluation and Research  
7    Center for Devices and Radiological Health  
8    Center for Devices and Radiological Health  
9    Center for Devices and Radiological Health

### Q.2 In which Year maximum number of emails were sent to doctors? 


In [228]:
emails['mail_sent_date'] = pd.to_datetime(emails['mail_sent_date'])

In [229]:
emails['mail_sent_date']=pd.to_datetime(emails['mail_sent_date']).dt.strftime('%Y')

In [230]:
count_years=emails.groupby('mail_sent_date').apply(lambda X:X['mail_sent_date'].count()).to_frame().reset_index()

In [231]:
count_years.head(15)

mail_sent_date   0
0            2006   1
1            2007  15
2            2008   8
3            2009  14
4            2010  25
5            2011  20
6            2012  16
7            2013  21
8            2014  32
9            2015  31
10           2016  25
11           2017  11
12           2018  23
13           2019  25
14           2020   5

###  Here we found that  maximum number of mails was sent  in 2014 that is 32 times

### Q.3 Write a query to get total fine paid per Office

### We have 2 data frames and the fines df contains fine details and the emails df contains office datasets

In [28]:
fine= fine.rename(columns={' cash_fine ': 'cash'})


In [29]:
fine= fine.rename(columns={' online_fine ': 'online'})

###  Here we renamed the 'cash_fine' and 'online_fine' columns to 'cash'  and 'online' due to error 'Data frame' object has no attribute

### We have to add a columns in fines df as total fine

In [23]:
# we need to check for null values
fine.isnull().sum()

org_indiv       0
first_name    203
last_name       0
city            0
state           0
category        1
cash            0
online          0
Total_fine      0
dtype: int64

In [22]:
emails.isnull().sum()

name_last          0
name_first         0
name_middle       91
mail_sent_date     0
office             0
dtype: int64

###  No relevant Null values

In [61]:
fine['online_replace']=fine['online'].str.replace('$','').str.replace(',','').str.replace('-', '0').str.replace('NA','0')


In [62]:
fine['online_replace']=pd.to_numeric(fine['online_replace'])

In [58]:
fine.online_replace

0          0.0
1          0.0
2        448.0
3        119.0
4          0.0
         ...  
10082      0.0
10083     65.0
10084     61.0
10085      0.0
10086     28.0
Name: online_replace, Length: 10087, dtype: float64

###  Now changed the data type to float 

In [63]:
# and also the cash column can be changed
fine['cash_replace']=fine['cash'].str.replace('$','').str.replace(',','').str.replace('-', '0').str.replace('NA','0')


In [65]:
fine['cash_replace']=pd.to_numeric(fine['cash_replace'])

In [66]:
fine.cash_replace

0        2625.0
1        1000.0
2           0.0
3           0.0
4        1800.0
          ...  
10082    1750.0
10083       0.0
10084       0.0
10085    2000.0
10086       0.0
Name: cash_replace, Length: 10087, dtype: float64

In [68]:
fine['Total']=fine['cash_replace']+fine['online_replace']

In [70]:
fine.head()

org_indiv first_name    last_name         city state  \
0  3-D MEDICAL SERVICES LLC     STEVEN  DEITELZWEIG  NEW ORLEANS    LA   
1          AA DOCTORS, INC.     AAKASH        AHUJA  PASO ROBLES    CA   
2    ABBO, LILIAN MARGARITA     LILIAN         ABBO        MIAMI    FL   
3    ABBO, LILIAN MARGARITA     LILIAN         ABBO        MIAMI    FL   
4    ABBO, LILIAN MARGARITA     LILIAN         ABBO        MIAMI    FL   

                  category        cash    online  cash_replace  \
0    Professional Advising  $2,625.00      $-           2625.0   
1        Expert-Led Forums  $1,000.00      $-           1000.0   
2  Business Related Travel       $-     $448.00            0.0   
3                    Meals       $-     $119.00            0.0   
4    Professional Advising  $1,800.00      $-           1800.0   

   online_replace   Total  
0             0.0  2625.0  
1             0.0  1000.0  
2           448.0   448.0  
3           119.0   119.0  
4             0.0  1800.0

### Well we have added Total column and our question is  Write a query to get total fine paid per Office¶

### We need to merge the 2 datasets to find out total fine paid per office

### Here 2 data frames have one common key  that is last name and first name has null values 

In [75]:
fine= fine.rename(columns={'last_name': 'LastName'})

In [77]:
emails= emails.rename(columns={'name_last': 'LastName'})

In [113]:
fine_email=pd.merge(fine,emails,on='LastName',how="inner")

In [100]:
fine_email.shape

(810, 15)

In [115]:
fine_email.head()

org_indiv first_name LastName         city state  \
0        AILEEN FELDMAN MD     AILEEN  FELDMAN  SOUTHINGTON    CT   
1        AILEEN FELDMAN MD     AILEEN  FELDMAN  SOUTHINGTON    CT   
2   FELDMAN, ARTHUR EDWARD     ARTHUR  FELDMAN   SOMERVILLE    NJ   
3   FELDMAN, ARTHUR EDWARD     ARTHUR  FELDMAN   SOMERVILLE    NJ   
4  FELDMAN, DARREN RICHARD     DARREN  FELDMAN     NEW YORK    NY   

                  category        cash    online  cash_replace  \
0  Business Related Travel       $-      $66.00            0.0   
1        Expert-Led Forums  $2,500.00      $-           2500.0   
2        Expert-Led Forums  $1,250.00      $-           1250.0   
3                    Meals       $-      $92.00            0.0   
4  Business Related Travel       $-     $260.00            0.0   

   online_replace   Total FirstName name_middle mail_sent_date  \
0            66.0    66.0      MARK          H.     2011-12-19   
1             0.0  2500.0      MARK          H.     2011-12-19   
2             0.0  1250.0      MARK          H.     2011-12-19   
3            92.0    92.0      MARK          H.     2011-12-19   
4           260.0   260.0      MARK          H.     2011-12-19   

                                       office  
0  Center for Devices and Radiological Health  
1  Center for Devices and Radiological Health  
2  Center for Devices and Radiological Health  
3  Center for Devices and Radiological Health  
4  Center for Devices and Radiological Health

### There are 810 records that have common last name within the 2 datasets 

### To find Totel fine per office 

In [116]:
total_per_office=fine_email.groupby('office').apply(lambda X:X['Total'].sum()).to_frame().reset_index()

In [117]:
total_per_office.columns=['office', 'Total Fine']

In [118]:
total_per_office.head(10)

office  Total Fine
0  Center for Biologics Evaluation and Research    309787.0
1    Center for Devices and Radiological Health   1001709.0
2       Center for Drug Evaluation and Research    571436.0
3                      New York District Office       780.0

### Here is the answer 

### Q.4 Write a query to get the highest total fine and lowest total fine paid per state

In [129]:
total_per_state=fine_email.groupby('state').apply(lambda X:X['Total'].sum()).to_frame().reset_index()

In [120]:
total_per_state.columns=['state','Total Fine']

In [126]:
total_per_state.max()

state             WV
Total Fine    294839
dtype: object

### WV state have the maximum Total fine

In [127]:
total_per_state.min()

state          AL
Total Fine    750
dtype: object

### AL state  have lowest Total fine

###  Q.5 Write a query to get the least 2 states that had to pay minimum 'online' fine 


In [143]:
total_per_state_min_online=fine_email.groupby('state').apply(lambda X:X['online_replace'].sum()).to_frame().reset_index()

In [144]:
total_per_state_min_online.columns=['state','online_fine']

In [148]:
total_per_state_min_online.sort_values(["online_fine"], axis=0,
                 ascending=True, inplace=True)

In [150]:
total_per_state_min_online.head(15)

state  online_fine
24    NE          0.0
23    ND          0.0
30    OK          0.0
27    NV          0.0
12    IN         70.0
10    ID         75.0
35    UT         82.0
9     IA        180.0
14    KY        601.0
0     AL        636.0
1     AR        663.0
5     CT        811.0
20    MO        977.0
16    MA       1030.0
6     DC       1042.0

### IN and ID state have least minumum value in online fines

###  Q.6 Write a SQL query to find the number of doctors that contributed to the state having highest 'cash' fine


In [206]:
import sqlite3

In [207]:
sql_connect = sqlite3.connect('fines.db')

In [208]:
c = sql_connect.cursor()


In [209]:
c.execute("CREATE TABLE IF NOT EXISTS fine ({})".format(' ,'.join(fine_email.columns)))
        

##  #due to some database errors I am writing just the SQL query to fetch the number of  doctors

### select count(LastName) from fine_email where state IN (select state from fine_email where cash_replace=(select max(cash_replace) from fine_email));

### Q.7 Get the top 5 average fine paid per state and category


In [172]:
top5_average_state=fine_email.groupby(['state']).apply(lambda X:X['Total'].mean()).to_frame().reset_index()

In [176]:
top5_average_state.columns=['state','avg_of_fine']

In [179]:
top5_average_state.head()

state  avg_of_fine
0    AL   803.000000
1    AR  2073.666667
2    AZ  3556.916667
3    CA  1969.000000
4    CO   658.700000

In [180]:
      top5_average_state.sort_values(["avg_of_fine"], axis=0,
                 ascending=False, inplace=True)          

In [181]:
top5_average_state.head()

state   avg_of_fine
25    NH  15248.500000
20    MO   9224.611111
24    NE   5285.000000
37    WA   5100.625000
34    TX   3732.139241

In [183]:
top5_average_state.round(0).head()

state  avg_of_fine
25    NH      15248.0
20    MO       9225.0
24    NE       5285.0
37    WA       5101.0
34    TX       3732.0

###  So we can say Top Average fine was paid by NH state 

### And in the case of category we can check 

In [184]:
top5_average_category=fine_email.groupby(['category']).apply(lambda X:X['Total'].mean()).to_frame().reset_index()

In [185]:
top5_average_category.columns=['category','avg_of_fine']

In [187]:
top5_average_category.sort_values(["avg_of_fine"], axis=0,
                 ascending=False, inplace=True)          

In [189]:
top5_average_category.round(0).head()

category  avg_of_fine
4                    Investigator-Initiated Research      37500.0
6                              Professional Advising       5510.0
2                                  Expert-Led Forums       3725.0
3  Infosys Sponsored Research initiated on or aft...       1071.0
0                            Business Related Travel        941.0

### And the  top category that have paid most average fine that is 37500 dollars and the category name is investigator-initiated Research


## Insights from these calculations

#### -> Found  data quality issue in case of No doctor_id or such primary key
#### -> Most average fines are paid doctors for Investigator Initiated Research and then proffessional Advising
#### -> NE,ND,OK and NV these are the state they didn't  paid online_fine